src/App.tsx

```diff
@@ -1,46 +1,35 @@
 import pDebounce from "p-debounce";
 import React, { Component, ChangeEvent } from "react";
 import randomize from "randomatic";
-import pkg from "../package.json";
 import "./App.css";
-import { CodapHelper as Codap, ISaveState } from "./lib/codap-helper";
+import { CodapHelper as Codap,  } from "./lib/codap-helper";
 import codapInterface from "./lib/CodapInterface";
 import { DB } from "./lib/db";
 import { DBSharedTable } from "./lib/db-types";
 import { DataContext, CodapItem, CodapRequest } from "./lib/types";
-
-const kPluginName = "Collaborative Data Sharing";
-const kVersion = pkg.version;
-const kInitialDimensions = {
-  width: 350,
-  height: 400
-};
-const kSharedDimensions = {
-  width: 350,
-  height: 200
-};
-
-const kShareIdLength = 6;
-
-const kNewSharedTable = "new-table";
-const kNewDataContextTitle = "Collaborative Table";
-
-interface IState extends ISaveState {
-  id: string;
-  availableDataContexts: DataContext[];
-  selectedDataContext: string;
-  personalDataLabel: string;
-  personalDataKey: string;
-  shareId?: string;
-  joinShareId: string;
-  isInProcessOfSharing: boolean;
-  showJoinShareError: boolean;
-}
+import { kInitialDimensions, kPluginName, kSharedDimensions, kVersion, kNewDataContextTitle,
+  kNewSharedTable, kShareIdLength } from "./constants";
+import { IState } from "./types";
+import { FirstPage } from "./ui-pages/first-page";
+import { JoinOtherTablePage } from "./ui-pages/join-other-table";
+import { AllowOthersToJoinPage } from "./ui-pages/allow-others-to-join-page";
+import { MergeTablePage } from "./ui-pages/merge-table-page";
+import { CreateNewTablePage } from "./ui-pages/create-new-table-page"
 
 let database: DB;
 
 export default class App extends Component {
 
+  constructor() {
+    super({});
+    this.setState = this.setState.bind(this);
+    this.handleJoinShareIdChange = this.handleJoinShareIdChange.bind(this);
+    this.handleDataLabelChange = this.handleDataLabelChange.bind(this);
+    this.joinShare = this.joinShare.bind(this);
+    this.handleDataContextChange = this.handleDataContextChange.bind(this);
+    this.initiateShare = this.initiateShare.bind(this);
```

Comment:
I was really confused when I first saw this, but now I see you're setting up functions you can pass as props to the page components. Instead of reassigning the value of class methods in the constructor, I think you should assign these to local variables inside of `renderFormPage`. You could also use `=>` notation, which might be a little more clear, though you'd have to specify arguments if you did. Then it would be something like:
`const setState = (state: Partial<IState>) => this.setState(state);`